File: python/src/aps_kit/runtime.py
"""Acceptance runtime. Mirrors specs/acceptance-generator.md runtime contract."""
from __future__ import annotations

from pathlib import Path
from typing import Any, Iterator

from .ir import Feature, Scenario, Step, load_ir
from .registry import Registry, default_registry


def executions_for(feature: Feature) -> Iterator[tuple[int, int]]:
    """Yield (scenario_index, example_index) pairs the runtime must execute.

    A scenario with no examples produces exactly one execution (example_index 0
    with an empty example object). A scenario with N example rows produces N
    executions.
    """
    for s_idx, scenario in enumerate(feature.scenarios):
        if not scenario.examples:
            yield s_idx, 0
        else:
            for e_idx in range(len(scenario.examples)):
                yield s_idx, e_idx


def _example_for(scenario: Scenario, example_index: int) -> dict[str, str]:
    if not scenario.examples:
        if example_index != 0:
            raise IndexError(
                f"scenario {scenario.name!r} has no examples; only example_index=0 is valid"
            )
        return {}
    return dict(scenario.examples[example_index])


def _run_step(step: Step, world: dict[str, Any], example: dict[str, str], registry: Registry) -> None:
    # Resolve the handler first so an unsupported step gives the most useful
    # error before we check anything else; then validate required example
    # values; then invoke. Same order in all four language runtimes.
    handler = registry.resolve(step.text)
    missing = [p for p in step.parameters if p not in example]
    if missing:
        raise AssertionError(
            f"step {step.text!r} references missing example values: {missing}"
        )
    handler(world, example)


def run_execution(
    ir_path: str | Path,
    scenario_index: int,
    example_index: int,
    registry: Registry | None = None,
) -> None:
    """Run a single scenario/example execution from the IR file at ``ir_path``.

    The runtime loads the IR fresh on each call so a mutator can swap the file
    between runs without re-importing the test module.
    """
    feature = load_ir(ir_path)
    if scenario_index < 0 or scenario_index >= len(feature.scenarios):
        raise AssertionError(
            f"scenario index {scenario_index} out of range; feature has "
            f"{len(feature.scenarios)} scenarios"
        )
    scenario = feature.scenarios[scenario_index]
    example = _example_for(scenario, example_index)
    reg = registry or default_registry
    world: dict[str, Any] = {}
    for step in feature.background:
        _run_step(step, world, example, reg)
    for step in scenario.steps:
        _run_step(step, world, example, reg)

File: typescript/src/runtime.ts
// Acceptance runtime. Mirrors specs/acceptance-generator.md runtime contract.
import type { Feature, Scenario, Step } from "./ir.js";
import { loadIR } from "./ir.js";
import { defaultRegistry, Registry } from "./registry.js";

export function executionsFor(feature: Feature): Array<[number, number]> {
  const out: Array<[number, number]> = [];
  feature.scenarios.forEach((scenario, sIdx) => {
    if (scenario.examples.length === 0) {
      out.push([sIdx, 0]);
    } else {
      scenario.examples.forEach((_, eIdx) => out.push([sIdx, eIdx]));
    }
  });
  return out;
}

function exampleFor(scenario: Scenario, exampleIndex: number): Record<string, string> {
  if (scenario.examples.length === 0) {
    if (exampleIndex !== 0) {
      throw new Error(
        `scenario ${scenario.name} has no examples; only exampleIndex=0 is valid`,
      );
    }
    return {};
  }
  return { ...scenario.examples[exampleIndex] };
}

function runStep(
  step: Step,
  world: Record<string, unknown>,
  example: Record<string, string>,
  registry: Registry,
): void {
  // Resolve the handler first so an unsupported step gives the most useful
  // error before we check anything else; then validate required example
  // values; then invoke. Same order in all four language runtimes.
  const handler = registry.resolve(step.text);
  const missing = step.parameters.filter((p) => !(p in example));
  if (missing.length > 0) {
    throw new Error(
      `step "${step.text}" references missing example values: ${missing.join(", ")}`,
    );
  }
  handler(world, example);
}

export function runExecution(
  irPath: string,
  scenarioIndex: number,
  exampleIndex: number,
  registry: Registry = defaultRegistry,
): void {
  const feature = loadIR(irPath);
  if (scenarioIndex < 0 || scenarioIndex >= feature.scenarios.length) {
    throw new Error(
      `scenario index ${scenarioIndex} out of range; feature has ${feature.scenarios.length} scenarios`,
    );
  }
  const scenario = feature.scenarios[scenarioIndex];
  const example = exampleFor(scenario, exampleIndex);
  const world: Record<string, unknown> = {};
  for (const step of feature.background) {
    runStep(step, world, example, registry);
  }
  for (const step of scenario.steps) {
    runStep(step, world, example, registry);
  }
}
